### PR TITLE
gh-55 Updated to include a local development environment via webpack-…

### PR DESCRIPTION
**Pull request by Spechal**
…dev-server, docker container and added nginx proxy to production image

gh-55 Updated to include a local development environment via webpack-dev-server, docker container and added nginx proxy to production image

### CODE REVIEW

**Comment by Spechal** [comment]
You can validate this PR by checking out the branch and following the instructions in the frontend/README.md file

File: frontend/README.md [hunk 1]
```diff
@@ -1,8 +1,30 @@
...
+This application can be run locally with webpack-dev-server or via a docker container.  Both methods below will result in a web site running in development mode available at http://localhost:8080
 
-Runs the app in the development mode.<br>
-Open [http://localhost:8080](http://localhost:8080) to view it in the browser.
+#### To run locally via webpack-dev-server:
+```
+$> npm install
+$> npm run dev-server
+```
+Visit http://localhost:8080 in your browser.
```
Comment (clementkng) [inline frontend/README.md]: This now opens up on port 3000 for me. Can you confirm the port?

File: frontend/README.md [hunk 2]
```diff
@@ -1,8 +1,30 @@
...
+#### To build a production Docker image
+```
+$> docker build --pull --no-cache --force-rm --quiet \
+  -t meep-frontend -f prod.Dockerfile .
+```
+
+#### To run the production image locally
+```
+$> docker run -it -p 8080:80 \
+  $(docker images -f label=project=meep -f label=component=frontend -q)
```
Comment (clementkng) [inline frontend/README.md]: Per our slack discussion, I still would push to have a platform agnostic command to run. This could be very helpful for first timers or other contributors who don't know docker that well.

File: frontend/dev.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+LABEL project="meep"
+LABEL component="frontend"
+
+RUN npm install -g --no-progress --ignore-optional webpack webpack-cli
+
+WORKDIR /tmp
+COPY package.json /tmp/
+RUN npm install --no-progress --ignore-optional
+
+WORKDIR /usr/src/app
```
Comment (clementkng) [inline frontend/dev.Dockerfile]: Is it necessary to move `WORKDIR` around?

File: frontend/prod.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,28 @@
+FROM node:alpine as build
```
Comment (clementkng) [inline frontend/prod.Dockerfile]: It would be nice to order the commands to correspond w/ the dev.Dockerfile for consistency.

File: frontend/nginx/nginx.conf [hunk 5]
```diff
@@ -0,0 +1,17 @@
+server {
+
+  listen 80;
+
+  location / {
+    root   /usr/share/nginx/html/public;
+    index  index.html index.htm;
```
Comment (clementkng) [inline frontend/nginx/nginx.conf]: Is there a reason `root` is defined inside the `location`? I've seen `nginx.conf` files use both, but just wondering.

Comment (Spechal) [reply frontend/nginx/nginx.conf]: It changes the root based on the location.

File: frontend/README.md (hunk 1 above)
Comment (Spechal) [reply frontend/README.md]: Docker maps the ports so you should still be hitting port 8080 for nginx to proxy to node

File: frontend/README.md (hunk 2 above)
Comment (Spechal) [reply frontend/README.md]: I suppose this wouldn't work on Windows. If you know the image ID to run, you can use it.  The command above finds out the ID for you based on the image's labels.

File: frontend/dev.Dockerfile (hunk 3 above)
Comment (Spechal) [reply frontend/dev.Dockerfile]: It depends on the subsequent actions.  In this case, I want to install the node packages in one place and then change the working directory for the app code.

File: frontend/prod.Dockerfile (hunk 4 above)
Comment (Spechal) [reply frontend/prod.Dockerfile]: We can, but they build two different kinds of images.  One for local development using webpack and another for production with nginx.

**Comment by Spechal** [comment]
> @Spechal I'm sorry for the extreme delay, and hope you had a nice 4th of July! I want to preface this by noting that I am still not an expert on Docker and Nginx, I just skimmed through the docs and played around w/ the PR a bit locally, so apologies in advance if this review has a few more questions than concrete feedback.
> 
> Additionally, it would give me peace of mind to have some sort of test for the Nginx configuration, but I believe that's beyond the scope of the PR. I'm pretty happy w/ this after my feedback. @galbwe @ASTPMeetup would you like to chime in?

We can add an nginx config check (`nginx -t`)to the build, but outside of that the image would have to be built and then a script run against it to verify integration.

**Comment by clementkng** [comment]
@Spechal so it seems they merged the other [docker PR](https://github.com/codeforkansascity/meep/pull/54), but there are still parts of this that do appear to be needed by this project ie the nginx reverse proxy. If you want to break out the parts this PR addresses into smaller PRs, I think others will have an easier time reviewing it. I can also try working it out as well if you're busy.

**Comment by Spechal** [comment]
I saw that too and was curious about it.  That branch could be merged into this one, but really, the work on that branch goes away with this one...

**Comment by clementkng** [comment]
I guess we could try some sort of manual rebase on the pulled code, and maybe split up the various problems this PR solves so it's easier to review.

File: frontend/package.json [hunk 6]
```diff
@@ -8,28 +8,37 @@
...
+    "react-scripts": "3.0.1",
+    "validator": "8.0.0",
+    "webpack": "4.34.0",
+    "webpack-cli": "3.3.4"
+  },
+  "optionalDependencies": {
+    "fsevents": "^2.0.7"
+  },
+  "devDependencies": {
+    "@babel/cli": "^7.4.4",
```
Comment (ASTPMeetup) [inline frontend/package.json]: Why did you move these babel packages to devDependencies?

Comment (Spechal) [reply frontend/package.json]: They aren't used or needed for production code to run, so they are development dependencies.  It helps keep production code cleaner and smaller.

File: frontend/webpack.config.js [hunk 7]
```diff
@@ -1,25 +1,31 @@
...
+  devtool: "cheap-module-eval-source-map",
   devServer: {
-    contentBase: path.join(__dirname, 'public'),
-    historyApiFallback: true
+    contentBase: path.join(__dirname, "public"),
+    historyApiFallback: true,
+    compress: true,
+    host: "0.0.0.0",
+    port: 3000,
+    disableHostCheck: true
```
[the remark this answers is not in the document]
Comment (Spechal) [reply frontend/webpack.config.js]: It bypasses host checking and is used mostly for Docker.

File: frontend/package.json (hunk 6 above)
Comment (ASTPMeetup) [reply frontend/package.json]: Babel complies react code to es5. its need in production for react to work.

Comment (Spechal) [reply frontend/package.json]: > Babel complies react code to es5. its need in production for react to work.

No, it doesn't.

https://stackoverflow.com/questions/40143357/do-you-put-babel-and-webpack-in-devdependencies-or-dependencies

File: frontend/prod.Dockerfile (hunk 4 above)
Comment (clementkng) [reply frontend/prod.Dockerfile]: I understand, but just for consistency's sake while someone's quickly skimming the dockerfiles, could you reorder this?

Comment (Spechal) [reply frontend/prod.Dockerfile]: > I understand, but just for consistency's sake while someone's quickly skimming the dockerfiles, could you reorder this?

Yes, I updated it to be near replicas.  The "as build" part is needed in the production file for the final image to use without having to build it every time; such as for an nginx config change.

Comment (clementkng) [reply frontend/prod.Dockerfile]: I guess what I mean is for them to be even closer, s.t. the order of the commands is the same when it can be i.e. `RUN webpack` after `RUN cp -a /tmp/node_modules /usr/src/app/` and `EXPOSE 80` after `COPY ./nginx/nginx.conf /etc/nginx/conf.d`. I'll admit this is a little nitpicky, but I think it would improve readability.

File: frontend/README.md (hunk 1 above)
Comment (clementkng) [reply frontend/README.md]: My point was actually regarding the instructions w/o docker, but I think it's probably a local issue that's causing the port to go to 3000.

**Comment by clementkng** [comment]
@ASTPMeetup Do you mean [this](https://github.com/codeforkansascity/meep/pull/67#issue-289301759)?

**Comment by Spechal** [comment]
@t-ketchem @galbwe Can we get some movement on this PR?

**Comment by ASTPMeetup** [comment]
@Spechal the docker-compose branch has not yet been successfully merged with master, blocking your PR. In addition, you currently have 2 conflicts with master.

**Comment by ASTPMeetup** [comment]
@Spechal last Monday, we discussed how we're handling PR submissions moving forward. You can find notes here -https://drive.google.com/drive/u/1/folders/1i9mHC7F3BDCu8quxVs44Jr3wpmZ7wkZa.

Every PR should include a detailed description of what it's trying to accomplish and screenshots of UI changes when applicable. Every Monday we plan on taking 10-15 minutes to discuss and review outstanding pull requests in person. From there, a volunteer will pull down your pull request branch and test your code to see if it works on their machine. After all question marks have been addressed and conflicts have been resolved, the PR will ready for merge consideration.

**Comment by Spechal** [comment]
I can fix the merge conflicts with master but why does your branch block my branch?

**Comment by ASTPMeetup** [comment]
@Spechal "I can fix the merge conflicts with master but why does your branch block my branch?"

Because your PR effects files currently being changed in the working docker-compose branch.

**Comment by Spechal** [comment]
But my branch was started way before yours and I own the ticket ... I don’t really care, I’m just confused as to how this logically works.
